### PR TITLE
fix: Scrolling of windows with winbar or native borders

### PR DESCRIPTION
**Pull request by fredizzimo**
## What kind of change does this PR introduce?
This fixes the scrolling of windows with winbar or native borders by only scrolling the inner region. 

The solution is quite hacky, see the new `neovide_unlink_border_highlights` option for example. But I think this is the best we can do with the current version of Neovim.

I also changed `Arc<Mutex<Line>>`  `<RC<RefCell<Line>>` because it's never accessed concurrently.

* Fixes https://github.com/neovide/neovide/issues/1550
* Fixes https://github.com/neovide/neovide/issues/1446

**NOTE:** This probably needs a bit more testing before merging.

## Did this PR introduce a breaking change? 
- No

### CODE REVIEW

File: src/bridge/events.rs [hunk 1]
```diff
@@ -523,13 +525,73 @@ fn parse_style(style_map: Value) -> Result<Style> {
...
+
+    for attribute in attributes {
+        if let (Value::String(name), value) = attribute {
+            match (name.as_str().unwrap(), value) {
+                ("kind", value) => {
+                    let kind_str = parse_string(value)?;
+                    match kind_str.as_str() {
+                        "ui" => kind = Some(HighlightKind::Ui),
+                        "syntax" => kind = Some(HighlightKind::Syntax),
+                        // The documentation says terminal but Neovim 0.9.4 sends term...
```
Comment (Kethku) [inline src/bridge/events.rs:552]: Is there an issue to link here?

File: src/bridge/events.rs [hunk 2]
```diff
@@ -523,13 +525,73 @@ fn parse_style(style_map: Value) -> Result<Style> {
...
+        "kind field not found in highlight info".to_string(),
+    ))?;
+    let ui_name = if kind == HighlightKind::Ui {
+        ui_name.ok_or(ParseError::Format(
+            "ui_name field not found in highlight info".to_string(),
+        ))?
+    } else {
+        String::default()
+    };
+    // hi_name can actually be absent for terminal, even though the documentation indicates otherwise
```
Comment (Kethku) [inline src/bridge/events.rs:578]: same here

**Comment by Kethku** [comment]
Looks good to me. Ship it ✅

**Comment by cshuaimin** [comment]
I tested this and it works for simple case like `set winbar=%F`, but not works with dropbar.nvim. When I change the highlight group, winbar scrolls with the content too, try `set winbar=%#Visual#%F`.

**Review by MultisampledNight — APPROVED** [review]

**Comment by cshuaimin** [comment]
That makes sense, thanks for the explanation. Float window border works well, great!

**Comment by ttytm** [comment]
I greatly appreciate the recent work of @fredizzimo. I want to say thanks at this point.

For this PR, I don't think it is ready for a merge. With the change, when using a winbar content that is more complex than something like `set winbar=%F`, the winbar content changes when e.g. opening a telescope picker. A change like this should not come with a downside for cases like @cshuaimin is mentioning or for users using things like dropbar. Here fix is not working anyway atm. But yeah, unfortunately, this again requires that upstream neovim allows for a proper fix.

E.g. current master opening a telescope file picker
Screenshot_20231213_151804

With the changes:
Screenshot_20231213_152130

**Comment by fredizzimo** [comment]
Hm.. this should be strictly better than main in all cases, winbars with custom highlight groups the bug should be the same as on main, and the content scrolls with the content like @cshuaimin explains,

I can't quite explain the different highlights in the screenshot above. But maybe the highlight groups are dynamically updated, and in that case the `neovide_unlink_border_highlights` hack does not work. @ttytm, can you test if setting that neovide option to false fixes it? 

If it does, there's no more hacks I can think of to get this work, and we will have to get a fix Neovim itself. Maybe we could merge this with the option set to false, to allow people to still have proper border and winbar scrolling with simple winbars and borders.

**Comment by ttytm** [comment]
> [...] If it does, there's no more hacks I can think of to get this work, and we will have to get a fix Neovim itself. Maybe we could merge this with the option set to false, to allow people to still have proper border and winbar scrolling with simple winbars and borders.

Good suggestion, I unfortunately cannot test it today anymore. I'll leave an update tomorrow.

**Comment by ttytm** [comment]
@fredizzimo the option has no effect on the mentioned behavior.

**Comment by fredizzimo** [comment]
@ttytm, I think I need a Neovim configuration, so that I can re-produce the problem myself.

Ideally following this format https://github.com/neovim/neovim/blob/master/contrib/minimal.lua

But if that is not possible, a complete `.config/nvim` directory with lazy as the plugin manager is also OK, since I can use `NVIM_APPNAME` to run the config without messing with my own.

**Comment by ttytm** [comment]
Hey @fredizzimo sorry bout the delay. I don't have the capacity atm to create a minimal repro. It's very unfortunate as I don't like to be a blocking factor. 

I can share a full nvim config where this happens with: https://github.com/tenxsoydev/nxvim.
Also, please feel free to ignore the mentioned objections.

**Comment by fredizzimo** [comment]
@ttytm, I tested your configuration, and I can repeat the problems with the winbar highlights getting brighter when telescope is opened. But I can also repeat that on the current master, so there must be something else going on here, mabye some earlier PR that broke it.

In any case, I will try to make this more robust by comparing the content instead of using highlight groups. We know how many lines are scrolled, and from that we see which lines stay in place. If it works, it should not take long to implement, so I'm putting this as draft until then.

**Comment by ttytm** [comment]
I'm discovering that it might not be a problem with Neovide but a problem with the config.
As I can reproduce the winbar problem when using nvim in Kitty on a mac. Somehow no the issue only appears in neovide with the PR but not in kitty at current master on Linux.

**Comment by fredizzimo** [comment]
I'm moving this out of draft again. The content based approach has too many special cases, and I'm not confident that it can be reliable enough. 

My idea was basically to check how many lines have stayed in place at the top and bottom, every time the screen is updated, and consider those borders or winbars, which are treated the same way in this PR. The biggest issue with that approach is duplicate lines, but also lines that changes due to highlight updates and stuff like that, or winbars with dynamic content. 

So instead I think we can go with what we have, perhaps even with `neovide_unlink_border_highlights` set to false by default, to make sure that it has no negative effects at all. And people that want the support can set that to true. Note that it needs to be set before the colorscheme is loaded.

But I'm leaving it enabled by default for now, until I get more feedback.

File: src/renderer/rendered_window.rs [hunk 3]
```diff
@@ -453,12 +507,37 @@ impl RenderedWindow {
...
+                                // The specification seems to indicate that kind should be UI and
+                                // then we only need to test ui_name. But at least for FloatTitle,
+                                // that is not the case, the kind is set to syntax and hi_name is
+                                // set.
+                                .map(|v| info.ui_name == *v || info.hi_name == *v)
+                                .any(|v| v)
+                            })
+                        })
+                    })
+                    .all(|v| v);
```
Comment (cshuaimin) [inline src/renderer/rendered_window.rs:534]:
```suggestion
                    .any(|v| v);
```

> lines with purly border highlight groups are considered borders.

I believe if we change this to "lines **contain** border highlight grouops are considered borders", then dropbar.nvim will work! Don't know if this has false positives. @fredizzimo what do you think?

Comment (cshuaimin) [reply src/renderer/rendered_window.rs:534]: However with this change, it seems that we need another way to ignore left and right borders.

Comment (fredizzimo) [reply src/renderer/rendered_window.rs:534]: That's a great idea!

We could use all to detect the borders and any to detect winbars. I will try that later today.

Comment (fredizzimo) [reply src/renderer/rendered_window.rs:532]: I implemented this, and it does seem to work. 

But I have only done a very quick test yet and it's a bit late for me to do more testing now.

**Comment by MultisampledNight** [comment]
Thank you! Choooooooooooo choooooooooooo :steam_locomotive:

**Pull request merged**